src/iop/hlrecovery.c

```diff
@@ -255,24 +282,227 @@ static inline int pos2plane(const int row, const int col, const uint32_t filters
+{
+  const int xmin = MAX(seg->xmin[id]-2, HLBORDER);
+  const int xmax = MIN(seg->xmax[id]+3, width - HLBORDER);
+  const int ymin = MAX(seg->ymin[id]-2, HLBORDER);
+  const int ymax = MIN(seg->ymax[id]+3, height - HLBORDER);
+  float max_distance = 0.0f;
+
+#ifdef _OPENMP
+  #pragma omp parallel for simd default(none) \
+  reduction(max : max_distance) \
+  dt_omp_firstprivate(distance, seg) \
+  dt_omp_sharedconst(width, xmin, xmax, ymin, ymax, id) \
+  schedule(simd:static) collapse(2)
+#endif
+  for(int row = ymin; row < ymax; row++)
+  {
+    for(int col = xmin; col < xmax; col++)
+    {
+      const size_t v = row * width + col;
+      if(id == seg->data[v])
+        max_distance = fmaxf(max_distance, distance[v]);
+    }
+  }  
+  return max_distance;
+}
+
+static float segment_attenuation(dt_iop_segmentation_t *seg, const int id, const int mode)
+{
+  const float attenuate[NUM_RECOVERY_MODES] = { 0.0f, 1.5f, 0.8f, 1.5f, 0.8f, 1.0f, 1.0f};
+  if(mode < RECOVERY_MODE_ADAPT)
+    return attenuate[mode];
+  else
+  {
+    const float maxdist = fmaxf(1.0f, seg->val1[id]);
+    return fminf(1.5f,  0.8f + (3.0f / maxdist)); 
+  }
+}
+
+static float segment_correction(dt_iop_segmentation_t *seg, const int id, const int mode, const int seg_border)
+{
+  const float correction = segment_attenuation(seg, id, mode);
+  return correction - 0.1f * (float)seg_border;
+}
+
+static void segment_gradients(const int width, const int height, float *restrict distance, float *restrict gradient, float *restrict tmp, const int mode, dt_iop_segmentation_t *seg, const int id, const int seg_border)
+{
+  const int xmin = MAX(seg->xmin[id]-1, HLBORDER);
+  const int xmax = MIN(seg->xmax[id]+2, width - HLBORDER);
+  const int ymin = MAX(seg->ymin[id]-1, HLBORDER);
+  const int ymax = MIN(seg->ymax[id]+2, height - HLBORDER);
+  const float attenuate = segment_attenuation(seg, id, mode);
+  const float strength = segment_correction(seg, id, mode, seg_border);
+
+  float maxdist = 1.5f;
+  while(maxdist < seg->val1[id])
+  {
+    calc_distance_ring(width, xmin, xmax, ymin, ymax, gradient, distance, attenuate, maxdist, seg, id);
+    maxdist += 1.5f;
+  }
+
+  if(maxdist > 4.0f)
+  {
+#ifdef _OPENMP
+  #pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(gradient, tmp) \
+  dt_omp_sharedconst(width, xmin, xmax, ymin, ymax) \
+  schedule(simd:static)
+#endif
+    for(int row = ymin; row < ymax; row++)
+    {
+      for(int col = xmin, s = row*width + col, d = (row-ymin)*(xmax-xmin) ; col < xmax; col++, s++, d++)
+        tmp[d] = gradient[s];
+    }
+
+    dt_box_mean(tmp, ymax-ymin, xmax-xmin, 1, MIN((int)maxdist, 15), 2);
+#ifdef _OPENMP
+  #pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(gradient, tmp, distance, seg) \
+  dt_omp_sharedconst(width, xmin, xmax, ymin, ymax, id) \
+  schedule(simd:static)
+#endif
+    for(int row = ymin; row < ymax; row++)
```

Comment:
size_t?

Reply:
As usual all your comments & request are good to me. Will fix Pretty soon